local-store/src/lib.rs

```diff
@@ -0,0 +1,280 @@
+			let local_txs: Vec<_> = ::serde_json::from_slice::<Vec<TransactionEntry>>(&val)
+				.map_err(Error::Json)?
+				.into_iter()
+				.filter_map(TransactionEntry::into_pending)
+				.collect();
+
+			Ok(local_txs)
+		} else {
+			Ok(Vec::new())
+		}
+	}
+
+	/// Update the entries in the database.
+	pub fn update(&self) -> Result<(), Error> {
+		let mut batch = self.db.transaction();
+
+		let local_entries: Vec<TransactionEntry> = self.node.pending_transactions()
+			.into_iter()
+			.map(Into::into)
+			.collect();
+
+		let local_json = ::serde_json::to_value(&local_entries).map_err(Error::Json)?;
+		let json_str = format!("{}", local_json);
+
+		batch.put_vec(self.col, LOCAL_TRANSACTIONS_KEY, json_str.into_bytes());
+		self.db.write(batch).map_err(Error::Database)
+	}
+}
+
+impl<T: NodeInfo> Drop for LocalDataStore<T> {
+	fn drop(&mut self) {
+		debug!(target: "local_store", "Updating node data store on shutdown.");
+
+		let _ = self.update();
+	}
+}
+
+#[cfg(test)]
+mod tests {
+	use super::{NodeInfo, LocalDataStore};
+
+	use std::sync::Arc;
+	use ethcore::transaction::{Transaction, Condition, PendingTransaction};
+	use ethkey::{Brain, Generator};
+
+	// we want to test: round-trip of good transactions.
+	// failure to roundtrip bad transactions (but that it doesn't panic)
+
+	struct Dummy(Vec<PendingTransaction>);
+	impl NodeInfo for Dummy {
+		fn pending_transactions(&self) -> Vec<PendingTransaction> { self.0.clone() }
+	}
+
+	#[test]
+	fn twice_empty() {
+		let db = Arc::new(::util::kvdb::in_memory(0));
+
+		{
+			let store = super::create(db.clone(), None, Dummy(vec![]));
+			assert_eq!(store.pending_transactions().unwrap(), vec![])
+		}
+
+		{
+			let store = super::create(db.clone(), None, Dummy(vec![]));
+			assert_eq!(store.pending_transactions().unwrap(), vec![])
+		}
+	}
+
+	#[test]
+	fn with_condition() {
+		let keypair = Brain::new("abcd".into()).generate().unwrap();
+		let transactions: Vec<_> = (0..10u64).map(|nonce| {
+			let mut tx = Transaction::default();
+			tx.nonce = nonce.into();
+
+			let signed = tx.sign(keypair.secret(), None);
+			let condition = match nonce {
+				5 => Some(Condition::Number(100_000)),
+				_ => None,
+			};
+
+			PendingTransaction::new(signed, condition)
+		}).collect();
+
+		let db = Arc::new(::util::kvdb::in_memory(0));
+
+		{
+			// nothing written yet, will write pending.
+			let store = super::create(db.clone(), None, Dummy(transactions.clone()));
+			assert_eq!(store.pending_transactions().unwrap(), vec![])
+		}
+		{
+			// pending written, will write nothing.
+			let store = super::create(db.clone(), None, Dummy(vec![]));
+			assert_eq!(store.pending_transactions().unwrap(), transactions)
+		}
+		{
+			// pending removed, will write nothing.
+			let store = super::create(db.clone(), None, Dummy(vec![]));
+			assert_eq!(store.pending_transactions().unwrap(), vec![])
+		}
+	}
+
+	#[test]
+	#[should_panic]
```

Comment:
So maybe it would be better to have `assert(expected != actual, "They aren't equal because loading skips all transactions with an invalid signature.")` instead? Should panic means that it can actually panic anywhere - imho it's suitable only for simple function calls not tests like this.